installdir/hldig.1.in
.TH hldig 1 "2018-04-04"
.\" NAME should be all caps, SECTION should be 1-8, maybe w/ subsection
.\" other parms are allowed: see man(7), man(1)
.SH NAME
hldig \- retrieve HTML documents for hl://Dig search engine
.SH SYNOPSIS
.B hldig
.I "[options]"
.SH "DESCRIPTION"
hldig retrieves HTML documents using the HTTP protocol and
gathers information from these documents which can later be
used to search these documents. This program can be referred to
as the search robot. 
.SH OPTIONS
.TP
.B \-
Get the list of URLs to start indexing from standard input. This will override the default parameter \fIstart_url\fR specified in the config file  and the file supplied to the \fI-m\fR option.
.TP
.B \-a
Use alternate work files. Tells hldig to append
.I .work
to database files, causing a second copy of the database to be
built. This allows the original files to be used by hlsearch
during the indexing run. 
.TP
.B \-c \fIconfigfile\fR
Use the specified
.I configfile
instead of the default. 
.TP
.B \-h \fImaxhops\fR
Restrict the dig to documents that are at most
.I maxhops
links away from the starting document. This only works if option \fI\-i\fR
is also given. 
.TP
.B \-i
Initial. Do not use any old databases. Old databases will be erased before runing the program.
.TP
.B \-m \fIfilename\fR
Minimal run. Only index the URLs given in the file \fIfilename\fR, ignoring all others. URLs in the file should be formatted one URL per line. 
.B \-s
Print statistics about the dig after completion. 
.TP
.B \-t
Create an ASCII version of the document database. This
database is easy to parse with other programs so that
information can be extracted from it for purposes other
than searching. One could gather some interesting statistics
from this database. 
.TS
cB cB 
c l .
Fieldname	Value
u	URL
t	Title
a	State 
	(0 normal, 1 not found, 2 not indexed, 3 obsolete)
m	Time of last modification reported by the server
s	Document Size in bytes
H	Excerpt of the document
h	Meta Description
l	Time of last rerievial
L	Count of links in the document or  of \fIoutgoing links\fR
b	Number of links to the document, also called 
	\fIincoming\fR links or \fIbacklinks\fR
c	Hop count of this document
g	Signature of this document 
	(used to detect duplicates)
e	E-Mail address to use for a notification from \fIhlnotify\fR
n	Date on which such notification is sent
S	Subject of the notfication message 
d	The text of Incoming links pointing to this document 
	(e.g. <a href="docURL">description</a>)
A	Anchors in the document (i.e. <A NAME=...)
.TE
.TP
.B \-u \fIusername:password\fR
Tells hldig to send the supplied username and password with
each HTTP request. The credentials will be encoded using the
\fI\'Basic\'\fR authentication method. There \fBHAS\fR to be a colon (:)
between the username and password. 
.TP
.B \-v
Verbose mode. This increases the verbosity of the program. Using more than 2 is probably only useful for debugging purposes. The default verbose mode (using only one -v) gives a nice progress report while digging. Please consult the section below on the exact format of the progress report.
.TP
.SH "FORMAT OF THE PROGRESS REPORT GIVEN IN VERBOSE MODE"
A line is shown for each URL, with 3 numbers before the URL and some symbols after the URL. The first number is the number of documents parsed so far, the second is the DocID for this document, and the third is the hop count of the document (number of hops from one of the start_url documents). Signification of the sybols printed after the url:
.IP 
\fI"*"\fR is printed for a link already visited
.IP 
\fI"+"\fR is printed for a new link just queued
.IP
\fI"-"\fR is output for a link rejected for any of a number of reasons. To find out what those reasons are, you need to run hldig with at least 3 \fI-v\fR options, i.e. \fI-vvv\fR. 
.TP
If there are no "*", "+" or "-" symbols after the URL, it doesn't mean the document was not parsed or was empty, but only that no links to other documents were found within it. With more verbose output, these symbols will get interspersed in several lines of debugging output. 
.TP
.SH "FILES"
.TP
.B @DEFAULT_CONFIG_FILE@
The default configuration file.
.B @DATABASE_DIR@/db.docdb
Stores data about each document (title, url, etc.).
.B @DATABASE_DIR@/db.words.db, @DATABASE_DIR@/db.words.db_weakcmpr
Record which documents each word occurs in.
.B @DATABASE_DIR@/db.excerpts
Stores start of each document to show context of matches.
.SH "SEE ALSO"
Please refer to the HTML pages (in the hldig-doc package)
.B /usr/share/doc/hldig-doc/html/index.html
and the manual pages
.I hldigconfig(8)
,
.I hlmerge(1)
,
.I hlnotify(1)
,
.I hlsearch(1)
and
.I rundig(1)
for a detailed description of hl://Dig and its commands.
.SH AUTHOR
This manual page was written by Christian Schwarz, modified by Stijn de Bekker. It is updated and maintained by Robert Ribnitz 
and based on the HTML documentation of ht://Dig and hl://Dig.


ChangeLog
2018-09-08  Johnny A. Solbu <johnny@solbu.net>

	* Don't pull external scripts. Serve it from our own server.
	  This also disables any tracking from the CDN provider.

2018-09-07  Johnny A. Solbu <johnny@solbu.net>

	* All tools should now be renamed

2018-09-03  Johnny A. Solbu <johnny@solbu.net>

	* Rename htfuzzy as hlfuzzy
	* Fix compile error on gcc 7, 8, and clang 6
	* Fix various compile warnings
	* Bump version: 0.3.0

2018-08-26  Johnny A. Solbu <johnny@solbu.net>

	* Release version 0.2.0
	* Image files doesn't need execute permission
	* Update TESTING.md with current build requirements
	* Remove non-existant benchmark folder from dist target:
	  git doesn't track empty folders
	* Some Makefile.am fixes


2018-04-04  Johnny A. Solbu <johnny@solbu.net>

	* Remove broken build test link
	* rename ht -> hl in topdir files
	* Complete rename in installdir
	* Autoreconf
	* Apparently several subscripts still use the VERSION file. Should
	fix that sometime
	* Rename mans in configure and make scripts.
	* Rename man files and their contents

2018-04-02  Johnny A. Solbu <johnny@solbu.net>

	* VERSION file is obsolete now that version is in configure.ac
	* Update pot file: copyright are assigned to «the hl://Dig group»

2018-03-27  andy5995 <andy400-dev@yahoo.com>

	* dbbench.cc:fix conversion warning This should fix the warning when `make dist` is run: dbbench.cc:269:20: warning: ISO C++ forbids converting a string
	constant to ‘char*’ [-Wpedantic]    char
	*progname = "dbbench problem...";

2018-03-28  Johnny A. Solbu <johnny@solbu.net>

	* bump version: 0.2.0
	* Add bug address to configure
	* Put version in the configure.ac file, where it belongs
	* Update build tests location
	* Rename url in the last files

2018-03-27  Johnny A. Solbu <johnny@solbu.net>

	* Rename urls in more files
	* Change MSGID_BUGS_ADDRESS, point to the issue tracker
	* Update ChangeLog.md with the new URLs
	* New project management: Rename URLs
	* Don't track compiled translations in revision controll
	* Remove unneeded preceding slash

2018-03-27  andy5995 <andy400-dev@yahoo.com>

	* change htdig.gif to logo.jpg (#56) After changes by @jamessouth, I found a few other areas I forgot to
	mention to him.

2018-03-15  jamessouth <jamesdanielsouth@gmail.com>

	* issue 56 new logo and modified html files

2018-03-15  Andy Alt <andy5995@users.noreply.github.com>

	* Changlog.md:add note about default install paths
	* Makefile.am:remove CVS dist-hook
	* CONTRIBUTING.md:add note about "skip ci"
	* TESTING.md:updated This reflects changes to the new install locations, which was a
	recent patch submitted by @Solbu
	* push regenerated build scripts and makefiles Regenerated due to previous commit
	* configure.ac:change path of www images When installdir/Makefile is run, it uses the value here to prefix
	urls to images in the installdir/html templates with /www/ To be accessible, a symlink can be put in the web root that points
	to that directory.

2018-03-15  Johnny A. Solbu <github@solbu.net>

	* Install things in the proper places (#90)
	  Install things in PREFIX/share/hldir, rather than /var/www

2018-03-12  Andy Alt <andy5995@users.noreply.github.com>

	* Merge pull request #101 from solbu/installer-fixes Installer fixes (Issue #99)
	* republish site from changes in bcdf311 This republishes the site using changes @jamessouth made in
	bcdf3110c78a308404431f3f3a24d33503d283a5 (#81)

2018-03-12  james south <jamessouth@users.noreply.github.com>

	* make sure home page validates (#100) (fixes #81)

2018-03-12  Johnny A. Solbu <johnny@solbu.net>

	* Add ABOUT-NLS to distributed files, indent traling backslash
	* Add missing gettext.h header

2018-03-11  andy5995 <andy400-dev@yahoo.com>

	* ABOUT-NLS:add to distribution
	* remove '#' (forgot to do this in a previous commit)
	* .travis.yml:drop '#' from hldig Freenode channel
	* simplecgen:refactor trim() function
	* list #hldig for the chat room, not ##hldig
	* .travis.yml: change failure and error notifications Only notify on_failure if there's a change Always notify on error
	* .travis.yml:change irc notification options
	* .travis.yml:change osx and Xcode version (#98)
	* .travis.yml:change osx and Xcode version
	* .travis.yml:add matrix closes #76

2018-03-11  Johnny A. Solbu <github@solbu.net>

	* Fix test dir (issue #91) (#97) / Thanks @solbu
	* Fix «make dist» Some tools where renamed at some point but the test commands using
	them where not updated
	* Update test/test_function.in with the renamed filenames

2018-03-09  Andy Alt <andy5995@users.noreply.github.com>

	* Merge pull request #92 from solbu/doc-changes Simplify translation instructions

2018-03-07  andy5995 <andy400-dev@yahoo.com>

	* translation:add ru to LINGUAS, generate gmo file (closes #68)
	* travis.yml:add osx (#76)
	* ne.po:remove empty file

2018-03-07  Svitlana Galianova <svitlana.galianova@gmail.com>

	* added russian translation

2018-03-07  Johnny A. Solbu <github@solbu.net>

	* Uninstall fix (#84) (closes #48)
	* Downgrade required gettext version, to match my system
	* Add Solbu to authors list

2018-03-06  andy5995 <andy400-dev@yahoo.com>

	* ChangeLog.md:note about html docs not being installed anymore

2018-03-06  Johnny A. Solbu <johnny@solbu.net>

	* Run autoreconf
	* Don't install documentaion in docdir RPM packagers prefer to control that dir using a «%doc» macro
	* Ignore .patch files To avoid committing patches by mistake, when one is also working on
	packaging

2018-01-21  andy5995 <andy400-dev@yahoo.com>

	* fix formatting
	* remove Slack, use IRC found some leftovers to remove (see previous commit)
	* remove Slack workspace from contact info IRC should be sufficient for this project.  http://webchat.freenode.net?channels=%23%23hldig&uio=d4

2018-01-15  andy5995 <andy400-dev@yahoo.com>

	* docs/README: explain how to view the html files

2018-01-11  Andy Alt <andy5995@users.noreply.github.com>

	* [#24]installdirs:update search result html, add css (#82)

hldig/hldig.cc
//
// hldig.cc
//
// hldig: Indexes the web sites specified in the config file
//        generating several databases to be used by htmerge
//
// Part of the hl://Dig package   <https://solbu.github.io/hldig/>
// Copyright (c) 2017 The hl://Dig Group
// For copyright details, see the file COPYING in your distribution
// or the GNU Library General Public License (LGPL) version 2 or later
// <http://www.gnu.org/copyleft/lgpl.html>
//
//  hl://Dig is a fork of ht://Dig <https://sourceforge.net/projects/htdig/>
//

#ifdef HAVE_CONFIG_H
#include "hlconfig.h"
#endif /* HAVE_CONFIG_H */

#include "Document.h"
#include "Retriever.h"
#include "StringList.h"
#include "hldig.h"
#include "defaults.h"
#include "HtURLCodec.h"
#include "WordContext.h"
#include "HtDateTime.h"
#include "HtURLRewriter.h"
#include "messages.h"

////////////////////////////
// For cookie jar
////////////////////////////
#include "HtCookieJar.h"
#include "HtCookieMemJar.h"
#include "HtCookieInFileJar.h"
#include "HtHTTP.h"
////////////////////////////

// If we have this, we probably want it.
#ifdef HAVE_GETOPT_H
#include <getopt.h>
#elif HAVE_GETOPT_LOCAL
#include <getopt_local.h>
#endif

#ifdef HAVE_STD
#include <iostream>
#ifdef HAVE_NAMESPACES
using namespace std;
#endif
#else
#include <iostream.h>
#endif /* HAVE_STD */

//
// Global variables
//
int debug = 0;
int report_statistics = 0;
DocumentDB docs;
HtRegexList limits;
HtRegexList limitsn;
FILE *urls_seen = NULL;
FILE *images_seen = NULL;
String configFile = DEFAULT_CONFIG_FILE;
String minimalFile = 0;
HtDateTime StartTime;
HtDateTime EndTime;

void usage ();
void reportError (const String& msg);


//
// Start of the program.
//
int
main (int ac, char **av)
{
  setlocale (LC_ALL, "");
  bindtextdomain (PACKAGE, LOCALEDIR);
  textdomain (PACKAGE);

  int c;
  extern char *optarg;
  String credentials;
  int initial = 0;
  int alt_work_area = 0;
  int create_text_database = 0;
  const char *max_hops = 0;

  // Cookie jar dynamic creation.
  HtCookieJar *_cookie_jar = new HtCookieMemJar ();     // new cookie jar
  if (_cookie_jar)
    HtHTTP::SetCookieJar (_cookie_jar);

//extern int yydebug;
//yydebug=1;

  //
  // Parse command line arguments
  //
  while ((c = getopt (ac, av, "lsm:c:vith:u:a")) != -1)
  {
    unsigned int pos;
    switch (c)
    {
    case 'c':
      configFile = optarg;
      break;
    case 'v':
      debug++;
      break;
    case 'i':
      initial++;
      break;
    case 't':
      create_text_database++;
      break;
    case 'h':
      max_hops = optarg;
      break;
    case 's':
      report_statistics++;
      break;
    case 'u':
      credentials = optarg;
      for (pos = 0; pos < strlen (optarg); pos++)
        optarg[pos] = '*';
      break;
    case 'a':
      alt_work_area++;
      break;
    case 'm':
      minimalFile = optarg;
      max_hops = "0";
      break;
    case '?':
      usage ();
    default:
      break;
    }
  }

  // Shows Start Time
  if (debug > 0)
    cout << _("hldig Start Time: ") << StartTime.GetAscTime () << endl;

  //
  // First set all the defaults and then read the specified config
  // file to override the defaults.
  //
  HtConfiguration *const config = HtConfiguration::config ();
  config->Defaults (&defaults[0]);
  if (access ((char *) configFile, R_OK) < 0)
  {
    reportError (form (_("Unable to find configuration file '%s'"),
                       configFile.get ()));
  }
  config->Read (configFile);

  // Warn user if any obsolete options are found in config file
  // For efficiency, check all fields here.  If different config
  // files are used for searching, obsolete options may remain
  const char *deprecatedOptions[] = {
    "heading_factor_1", "heading_factor_2", "heading_factor_3",
    "heading_factor_4", "heading_factor_5", "heading_factor_6",
    "modification_time_is_now", "pdf_parser", "translate_amp",
    "translate_lt_gt", "translate_quot", "uncoded_db_compatible",
    ""                          // empty terminator
  };
  const char **option;
  for (option = deprecatedOptions; **option; option++)
  {
    if (!config->Find (*option).empty ())
      cout << _("Warning: Configuration option ") << *option <<
        _(" is no longer supported\n");
  }

  if (config->Find (_("locale")).empty () && debug > 0)
    cout << _("Warning: unknown locale!\n");

  if (max_hops)
  {
    config->Add (_("max_hop_count"), max_hops);
  }

  // Set up credentials for this run
  if (credentials.length ())
    config->Add (_("authorization"), credentials);

  //
  // Check url_part_aliases and common_url_parts for
  // errors.
  String url_part_errors = HtURLCodec::instance ()->ErrMsg ();

  if (url_part_errors.length () != 0)
    reportError (form (_("Invalid url_part_aliases or common_url_parts: %s"),
                       url_part_errors.get ()));

  //
  // Check url_rewrite_rules for errors.
  String url_rewrite_rules = HtURLRewriter::instance ()->ErrMsg ();

  if (url_rewrite_rules.length () != 0)
    reportError (form (_("Invalid url_rewrite_rules: %s"),
                       url_rewrite_rules.get ()));

  //
  // If indicated, change the database file names to have the .work
  // extension
  //
  if (alt_work_area != 0)
  {
    String configValue = config->Find ("doc_db");

    if (configValue.length () != 0)
    {
      configValue << ".work";
      config->Add ("doc_db", configValue);
    }

    configValue = config->Find ("word_db");
    if (configValue.length () != 0)
    {
      configValue << ".work";
      config->Add ("word_db", configValue);
    }

    configValue = config->Find ("doc_index");
    if (configValue.length () != 0)
    {
      configValue << ".work";
      config->Add ("doc_index", configValue);
    }

    configValue = config->Find ("doc_excerpt");
    if (configValue.length () != 0)
    {
      configValue << ".work";
      config->Add ("doc_excerpt", configValue);
    }

    configValue = config->Find ("md5_db");
    if (configValue.length () != 0)
    {
      configValue << ".work";
      config->Add ("md5_db", configValue);
    }
  }

  // Imports the cookies file
  const String CookiesInputFile = config->Find ("cookies_input_file");
  if (CookiesInputFile.length ())
  {
    if (debug > 0)
      cout << _("Importing Cookies input file ") << CookiesInputFile << endl;
    int result;
    HtCookieJar::SetDebugLevel (debug); // Set the debug level
    HtCookieInFileJar *cookie_file =
      new HtCookieInFileJar (CookiesInputFile, result);
    if (cookie_file)
    {
      if (!result)
      {
        if (debug > 0)
          cookie_file->ShowSummary ();
        delete _cookie_jar;     // Deletes previous cookie jar
        _cookie_jar = (HtCookieJar *) cookie_file;      // set the imported one
        HtHTTP::SetCookieJar (_cookie_jar);     // and set the new HTTP jar
      }
      else if (debug > 0)
        cout << _("Warning: Import failed! (") << CookiesInputFile << ")" <<
          endl;
    }
    else
      reportError (form (_("Unable to load cookies file '%s' in memory"),
                         CookiesInputFile.get ()));
  }

  //
  // If needed, we will create a list of every URL we come across.
  //
  if (config->Boolean ("create_url_list"))
  {
    const String filename = config->Find ("url_list");
    urls_seen = fopen (filename, initial ? "w" : "a");
    if (urls_seen == 0)
    {
      reportError (form ("Unable to create URL file '%s'", filename.get ()));
    }
  }

  //
  // If needed, we will create a list of every image we come across.
  //
  if (config->Boolean ("create_image_list"))
  {
    const String filename = config->Find ("image_list");
    images_seen = fopen (filename, initial ? "w" : "a");
    if (images_seen == 0)
    {
      reportError (form (_("Unable to create images file '%s'"),
                         filename.get ()));
    }
  }

  //
  // Set up the limits list
  //
  StringList l (config->Find ("limit_urls_to"), " \t");
  limits.setEscaped (l, config->Boolean ("case_sensitive"));
  l.Destroy ();

  l.Create (config->Find ("limit_normalized"), " \t");
  limitsn.setEscaped (l, config->Boolean ("case_sensitive"));
  l.Destroy ();

  //
  // Open the document database
  //
  const String filename = config->Find ("doc_db");
  if (initial)
    unlink (filename);

  const String index_filename = config->Find ("doc_index");
  if (initial)
    unlink (index_filename);

  const String head_filename = config->Find ("doc_excerpt");
  if (initial)
    unlink (head_filename);

  if (docs.Open (filename, index_filename, head_filename) < 0)
  {
    reportError (form (_("Unable to open/create document database '%s'"),
                       filename.get ()));
  }

  const String word_filename = config->Find ("word_db");
  if (initial)
  {
    unlink (word_filename);
    unlink ((word_filename + "_weakcmpr").get ());

    // Remove "duplicate detection" database
    unlink (config->Find ("md5_db"));

    // using  -i,  also ignore seen-but-not-processed URLs from last pass
    unlink (config->Find ("url_log"));
  }

  // Initialize htword
  WordContext::Initialize (*config);

  // Create the Retriever object which we will use to parse all the
  // HTML files.
  // In case this is just an update dig, we will add all existing
  // URLs?
  //
  Retriever retriever (Retriever_logUrl);
  if (minimalFile.length () == 0)
  {
    List *list = docs.URLs ();
    retriever.Initial (*list);
    delete list;

    // Add start_url to the initial list of the retriever.
    // Don't check a URL twice!
    // Beware order is important, if this bugs you could change
    // previous line retriever.Initial(*list, 0) to Initial(*list,1)
    retriever.Initial (config->Find ("start_url"), 1);
  }

  // Handle list of URLs given in a file (stdin, if "-") specified as
  // argument to -m or as an optional trailing argument.

  // FIXME: this string needs changing for use with getttext //
  if (optind < ac)
  {
    if (debug)
      if (minimalFile.length () != 0)
        cout << "Warning: argument " << av[optind]
          << " overrides -m " << minimalFile << endl;
    minimalFile = av[optind];
  }
  if (strcmp (minimalFile.get (), "-") == 0)
  {
    String str;
    // Why not combine this with the code below, with  input = stdin ?
    while (!cin.eof ())
    {
      cin >> str;
      str.chop ("\r\n");        // (Why "\r\n" here and "\r\n\t " below?)
      if (str.length () > 0)
        retriever.Initial (str, 1);
    }
  }
  else if (minimalFile.length () != 0)
  {
    FILE *input = fopen (minimalFile.get (), "r");
    char buffer[1000];

    if (input)
    {
      while (fgets (buffer, sizeof (buffer), input))
      {
        String str (buffer);
        str.chop ("\r\n\t ");
        if (str.length () > 0)
          retriever.Initial (str, 1);
      }
      fclose (input);
    }
    else
    {
      fprintf (stderr, _("Could not open argument '%s' of flag -m\n"), &minimalFile[0]);
      exit (1);
    }
  }

  //
  // Go do it!
  //
  retriever.Start ();

  //
  // All done with parsing.
  //

  //
  // If the user so wants, create a text version of the document database.
  //

  if (create_text_database)
  {
    const String doc_list = config->Find ("doc_list");
    if (initial)
      unlink (doc_list);
    docs.DumpDB (doc_list);
    const String word_dump = config->Find ("word_dump");
    if (initial)
      unlink (word_dump);
    HtWordList words (*config);
    if (words.Open (config->Find ("word_db"), O_RDONLY) == OK)
    {
      words.Dump (word_dump);
    }
  }

  //
  // Cleanup
  //
  if (urls_seen)
    fclose (urls_seen);
  if (images_seen)
    fclose (images_seen);

  //
  // If needed, report some statistics
  //
  if (report_statistics)
  {
    retriever.ReportStatistics ("hldig");
  }

  // Shows End Time
  if (debug > 0)
  {
    EndTime.SettoNow ();
    cout << _("hldig End Time: ") << EndTime.GetAscTime () << endl;
  }

  if (_cookie_jar)
    delete _cookie_jar;
}


//
// Display usage information for the hldig program
//
void
usage ()
{
  Usage help;
  cout << _("usage:");
  cout << " hldig [-v][-i][-c configfile][-t][-m minimalfile]\n";
  printf (_("This program is part of hl://Dig %s\n\n"), VERSION);
  cout << _("Options:\n");

  help.verbose ();

  printf (_("\
 -i\tInitial.  Do not use any old databases.  This is\n\
\taccomplished by first erasing the databases.\n\n"));

  help.config ();

  printf (_("\
 -t\tCreate an ASCII version of the document database.\n\
\tThis database is easy to parse with other programs so\n\
\tthat information can be extracted from it.\n\n"));

  printf (_("\
 -h\thop count\n\
\tLimit the stored documents to those which are at\n\
\tmost hopcount links away from the start URL.\n\n"));

  printf (_("\
 -s\tReport statistics after completion.\n\n"));

  printf (_("\
 -u\tusername:password\n\
\tTells hldig to send the supplied username and\n\
\tpassword with each HTTP request.  The credentials\n\
\twill be encoded using the 'Basic' authentication scheme.\n\
\tThere *HAS* to be a colon (:) between the username\n\
\tand password.\n\n"));

  help.alternate_common();
  printf (_("\
\tTells hldig to append .work to database files, causing\n\
\ta second copy of the database to be built.  This allows\n\
\tthe original files to be used by hlsearch during the\n\
\tindexing run.\n\n"));

  printf (_("\
 -m\tminimalfile (or just a file name at end of arguments)\n\
\tTells hldig to read URLs from the supplied file and index\n\
\tthem in place of (or in addition to) the existing URLs in\n\
\tthe database and the start_url.  With the -m, only the\n\
\tURLs specified are added to the database.  A file name of\n\
\t'-' indicates the standard input.\n\n"));

  exit (0);
}

//
// Report an error and die
//
void
reportError (const String& msg)
{
  cout << "hldig: " << msg << "\n\n";
  exit (1);
}
